content/python/concepts/files/terms/context-managers/context-managers.md

```diff
@@ -0,0 +1,139 @@
+---
+Title: 'Context Managers and Files'
+Description: 'This document introduces context managers and in particular how they can be used to work with files. It details the two approaches, class-based and decorator-based, thanks to which the user can customize this experience to fit the needs of a given use case.'
+Subjects:
+  - 'Computer Science'
+  - 'Data Science'
+Tags:
+  - 'Best Practices'
+  - 'Classes'
+  - 'Decorators'
+  - 'Files'
+CatalogContent:
+  - 'learn-python-3'
+  - 'paths/computer-science'
+---
+
+## Definition of a context managers
+
+One might want to perform some operation within a certain overarching context, be it in the form of a time counter or a database connection or the use of a [file](https://www.codecademy.com/resources/docs/python/files). This is what **context managers** allow to do by implementing a **runtime context**, within which can be perform the wanted code and that closes automatically just after, sparing to the user the hassle of doing so. Regardless of the many different use cases a context manager can be used for, the **context manager protocol** always articulates those four functionalities: open the resource, implement the runtime context, handle the possible [exceptions](https://www.codecademy.com/resources/docs/python/errors), close the resource. The user input finds its place once the runtime context is started.
+
+## Manipulate files with context managers
+
+The `open()` function is the standard context manager to open, perform operations on and close files.
+
+```py
+# this code will print the first four lines of the poem "L'albatros"
+with open('albatros.txt','r') as poem:
+  for i in range(4):
+    print(poem.readline())
+
+# this code will however create an error as we are already outside of the indented block and therefore no longer within the context that had been setup
+print(albatros.readline())
+```
+
+Output:
+
+```md
+Souvent, pour s'amuser, les hommes d'équipage
+
+Prennent des albatros, vastes oiseaux des mers,
+
+Qui suivent, indolents compagnons de voyage,
+
+Le navire glissant sur les gouffres amers.
+```
+
+## Customized context managers for files
+
+While `open()` is at the core of context managers working with files, it is still possible to implement context managers with additional capabilities, for example, to personalize files written using them. It can be done via two approaches: the **class-based approach**, and the **decorator-based approach**. Follow the links for more on [classes](https://www.codecademy.com/resources/docs/python/classes) and [decorators](https://www.codecademy.com/resources/docs/python/decorators).
```

Comment:
```suggestion
## Customized Context Managers

While `open()` is at the core of using context managers with files, it is still possible to implement context managers with additional capabilities, e.g., personalizing files written using them. It can be done via two approaches:

- [Class](https://www.codecademy.com/resources/docs/python/classes)-based approach
- [Decorator](https://www.codecademy.com/resources/docs/python/decorators)-based approach
```